### PR TITLE
Document profile files

### LINKED ISSUE #1512: Add documentation for new --profile CLI option
this feature will be in 24.2.0 (next upcoming release) - aiming for week of 23 Feb

Draft documentation below - changes I would make 
* link to the actual profiles rather than duplicating the settings in the docs (that way we don't have to update if the profiles get tweaked) https://github.com/hyperledger/besu/blob/8b64023a121ea996ef60e4b7e2299c5807683f90/config/src/main/resources/profiles/minimalist-staker.toml
* enterprise or private - https://github.com/hyperledger/besu/blob/8b64023a121ea996ef60e4b7e2299c5807683f90/config/src/main/resources/profiles/enterprise-private.toml
* or just the directory https://github.com/hyperledger/besu/tree/8b64023a121ea996ef60e4b7e2299c5807683f90/config/src/main/resources/profiles

this issue created from besu issue https://github.com/hyperledger/besu/issues/6409
------------

# User Documentation

## Introduction to --profile CLI Option in Besu

Besu introduces a new feature, the `--profile` CLI option. This feature allows users to load existing TOML configuration files from resources, thereby enabling the override of default options. 

### How to Use the --profile CLI Option

To use the `--profile` CLI option, you need to specify it when running Besu, along with the name of the profile you want to use. For example:

```bash
besu --config-file=config.toml --profile=STAKER
```

In this case, `config.toml` is the user-provided configuration file, and `STAKER` is the pre-configured profile containing custom settings.

### Configuration Resolution

The configuration provider resolves settings in a cascading manner. It prioritizes settings in the following order:

1. Environment variables
2. Configuration file (specified by the `--config-file` option)
3. Profile file (specified by the `--profile` option)
4. Default values (used if no other configuration source is available)

For instance, if a setting is not found in the environment variables, Besu will look for it in the `config.toml` file. If the setting is not found in the `config.toml` file, Besu will then look for it in the `STAKER.toml` file. If the setting is not found in the `STAKER.toml` file, Besu will use the universal default value for that setting.

### Available Profiles

The `--profile` flag accepts the following profiles:

- `minimalist-staker`
- `staker`
- `private` and `enterprise` (both alias to the same profile)

Any configuration options explicitly set in the config file or command line will overwrite any of the defaults set via the profile.

## Minimalist Staker Profile

For stakers who want to maximize their hardware value but don't want to serve full sets of data to their peers, Besu offers a `minimalist_staker` profile. This profile implements the following defaults:

- Sync mode: `CHECKPOINT`
- Data-storage-format: `BONSAI`
- bonsai-historical-block-limit: `128`
- Max-peers: `25`

To use the `minimalist_staker` profile, run Besu with the `--profile` option set to `minimalist_staker`:

```bash
besu --profile=minimalist_staker
```

Please note that any configuration options explicitly set in the config file or command line will overwrite the defaults set via the `minimalist_staker` profile.

## Staker Profile

For stakers who want to maximize their hardware value and also want to serve full sets of data to their peers, Besu offers a `staker` profile. 

Please note that any configuration options explicitly set in the config file or command line will overwrite the defaults set via the `staker` profile.

## Enterprise/Private Profile

For private network operators who want to minimize unexpected confusion by setting sensible defaults, Besu offers an `enterprise` or `private` profile. This profile is designed to handle specific use-cases for private network operators.

To use the `enterprise` or `private` profile, run Besu with the `--profile` option set to `enterprise` or `private`:

```bash
besu --profile=enterprise
```
or
```bash
besu --profile=private
```

The `enterprise` or `private` profile implements the following defaults:

- Sync-mode: `FAST`
- `-sync-min-peers=1`
- `--remote-connections-limit-enabled=false`
- Tx-pool: `Legacy / Sequential`
- `--tx-pool-no-local-priority`
- `--tx-pool-limit-by-account-percentage` (set to a value larger than 4 transactions)

Please note that any configuration options explicitly set in the config file or command line will overwrite the defaults set via the `enterprise` or `private` profile.

### PR DESCRIPTION
**Pull request by alexandratran**
Document `--profile` CLI option and add how-to page on using profile files. Slight updates to the "Use a configuration file" page. Fixes #1512.

Previews:
- `--profile` documentation: https://besu-docs-qvq914ru4-hyperledger.vercel.app/development/public-networks/reference/cli/options#profile
- Use a configuration file: https://besu-docs-qvq914ru4-hyperledger.vercel.app/development/public-networks/how-to/use-configuration-file
- Use a profile file: https://besu-docs-qvq914ru4-hyperledger.vercel.app/development/public-networks/how-to/use-configuration-file/profile-file

### CODE REVIEW

File: docs/public-networks/how-to/use-configuration-file/profile-file.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+---
+sidebar_position: 1
+---
+
+# Use a profile file
+
+You can load TOML configuration files containing custom settings for specific user profiles using
+the [`--profile`](../../reference/cli/options.md#profile) CLI option.
```
Comment (non-fungible-nelson) [inline docs/public-networks/how-to/use-configuration-file/profile-file.md:8]: Maybe something like - Besu provides some profiles for some common use-cases.

**Review by non-fungible-nelson — APPROVED** [review]
LGTM - small nit, not a blocker

File: docs/public-networks/how-to/use-configuration-file/profile-file.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+- [Enterprise/Private profile](#enterpriseprivate-profile)
+
+## Minimalist staker profile
+
+For stakers who want to maximize their hardware value but don't want to serve full sets of data to
+their peers, Besu provides a minimalist staker profile.
+See the
+[minimalist staker profile file](https://github.com/hyperledger/besu/blob/8b64023a121ea996ef60e4b7e2299c5807683f90/config/src/main/resources/profiles/minimalist-staker.toml)
+for the custom settings.
+
```
Comment (bgravenorst) [inline docs/public-networks/how-to/use-configuration-file/profile-file.md:35]: I wonder if there's a way in which we could actually display contents of the toml file?

For example: https://github.com/hyperledger/besu/blob/8b64023a121ea996ef60e4b7e2299c5807683f90/config/src/main/resources/profiles/enterprise-private.toml

So users can see what the values are, and minimise maintenance for us if the file changes.

**Comment by bgravenorst** [comment]
Do the various profiles affect the performance of the Besu node in any way? May be worth mentioning.

**Review by bgravenorst — APPROVED** [review]
1 minor comment and a couple of questions.

File: docs/public-networks/how-to/use-configuration-file/profile-file.md (hunk 2 above)
Comment (alexandratran) [reply docs/public-networks/how-to/use-configuration-file/profile-file.md:35]: I can't quite figure out an easy way to do this in Docusaurus. I tried [this solution](https://stackoverflow.com/questions/28338017/is-there-a-way-to-embed-github-code-into-an-iframe) and it caused some errors. Maybe something to solve in the future as this need arises again.

File: docs/public-networks/how-to/use-configuration-file/profile-file.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+This profile is designed to handle specific use cases for private network operators.
+See the
+[enterprise/private profile file](https://github.com/hyperledger/besu/blob/8b64023a121ea996ef60e4b7e2299c5807683f90/config/src/main/resources/profiles/enterprise-private.toml)
+for the custom settings.
+
+To use the enterprise/private profile, run Besu with
+[`--profile`](../../reference/cli/options.md#profile) set to `enterprise` or `private`:
+
+```bash
+besu --profile=staker
```
Comment (macfarla) [inline docs/public-networks/how-to/use-configuration-file/profile-file.md:70]:
```suggestion
besu --profile=enterprise
```

File: docs/public-networks/reference/cli/options.md [hunk 4]
```diff
@@ -2672,6 +2672,46 @@ p2p-port="1789"
...
+<TabItem value="Configuration file">
+
+```bash
+profile="staker"
+```
+
+</TabItem>
+</Tabs>
+
+Loads a TOML configuration file containing custom settings for a specific user profile.
```
[the remark this answers is not in the document]
Comment (macfarla) [reply docs/public-networks/reference/cli/options.md:2708]: idk if "pre-configured" is quite the right word - because you can't change the configuration, not in that TOML file anyway

[the remark this answers is not in the document]
Comment (macfarla) [reply docs/public-networks/reference/cli/options.md:2708]: maybe more like "These pre-configured TOML files are available, and you can also still build your own configuration files as well."

**Review by macfarla — COMMENTED** [review]
"profile file" always sounds weird in my head but I can't think how you would reword it - other than just using "profile"

File: docs/public-networks/how-to/use-configuration-file/profile-file.md (hunk 2 above)
Comment (bgravenorst) [reply docs/public-networks/how-to/use-configuration-file/profile-file.md:35]: Yeah GitHub probably doesn't allow content to be embedded in iFrames.

**Comment by macfarla** [comment]
> Do the various profiles affect the performance of the Besu node in any way? May be worth mentioning. Also, what problem is `--profiles` solving?

using a profile won't change performance (although the settings within the profile might do - that's kind of the point of them).

what problem are they solving? trying to make besu easier to "get started" with - there's a bit more context here https://wiki.hyperledger.org/display/BESU/Change+to+Besu+Defaults+-+Q1+2024

**Comment by alexandratran** [comment]
> "profile file" always sounds weird in my head but I can't think how you would reword it - other than just using "profile"

@macfarla I updated the wording to just "profile" and it sounds better.
